kotlinpoet/src/commonMain/kotlin/com/squareup/kotlinpoet/NameAllocator.kt

```diff
@@ -77,7 +77,39 @@ public class NameAllocator private constructor(
   private val allocatedNames: MutableSet<String>,
   private val tagToName: MutableMap<Any, String>,
 ) {
-  public constructor() : this(mutableSetOf(), mutableMapOf())
+  public constructor() : this(preAllocateKeywords = true)
+
+  /**
+   * @param preAllocateKeywords If true, all Kotlin keywords will be pre-allocated and suffixed with
+   * underscores to avoid using them as identifiers:
+   *
+   * ```kotlin
+   * val nameAllocator = NameAllocator(preAllocateKeywords = true)
+   * println(nameAllocator.newName("when")) // prints "when_"
+   * ```
+   *
+   * If false, keywords will not get any special treatment:
+   *
+   * ```kotlin
+   * val nameAllocator = NameAllocator(preAllocateKeywords = false)
+   * println(nameAllocator.newName("when")) // prints "when"
+   * ```
+   *
+   * Note that you can use the `%N` placeholder when emitting a name produced by [NameAllocator] to
+   * ensure it's properly escaped for use as an identifier:
+   *
+   * ```kotlin
+   * val nameAllocator = NameAllocator(preAllocateKeywords = false)
+   * println(CodeBlock.of("%N", nameAllocator.newName("when"))) // prints "`when`"
+   * ```
+   *
+   * The default behaviour of [NameAllocator] is to pre-allocate keywords - this is the behaviour you'll
+   * get when using the no-arg constructor.
+   */
+  public constructor(preAllocateKeywords: Boolean) : this(
```

Comment:
@swankjesse I decided to go with this naming after re-reading the docs for NameAllocator, I think it's more consistent with the terminology we use there. Here's our original diff, let me know if you prefer the name we originally came up with:

```diff
diff --git a/kotlinpoet/src/commonMain/kotlin/com/squareup/kotlinpoet/NameAllocator.kt b/kotlinpoet/src/commonMain/kotlin/com/squareup/kotlinpoet/NameAllocator.kt
index cda10042..514fb8c4 100644
--- a/kotlinpoet/src/commonMain/kotlin/com/squareup/kotlinpoet/NameAllocator.kt
+++ b/kotlinpoet/src/commonMain/kotlin/com/squareup/kotlinpoet/NameAllocator.kt
@@ -77,7 +77,20 @@ public class NameAllocator private constructor(
   private val allocatedNames: MutableSet<String>,
   private val tagToName: MutableMap<Any, String>,
 ) {
-  public constructor() : this(mutableSetOf(), mutableMapOf())
+  public constructor() : this(
+    allocatedNames = KEYWORDS.toMutableSet(),
+    tagToName = mutableMapOf(),
+  )
+
+  /**
+   * @param allowKeywords true to return keywords like "data" and "value" from [newName], where they'll potentially
+   *     be emitted with backticks when used to generate source code. The default is false, which will cause programs
+   *     to generate APIs like "data_" and "value_" which do not need to be escaped with backticks.
+   */
+  public constructor(allowKeywords: Boolean) : this(
+    allocatedNames = if (allowKeywords) mutableSetOf<String>() else KEYWORDS.toMutableSet(),
+    tagToName = mutableMapOf(),
+  )
 
   /**
    * Return a new name using `suggestion` that will not be a Java identifier or clash with other
@@ -89,7 +102,7 @@ public class NameAllocator private constructor(
     tag: Any = UUID.randomUUID().toString(),
   ): String {
     var result = toJavaIdentifier(suggestion)
-    while (result.isKeyword || !allocatedNames.add(result)) {
+    while (!allocatedNames.add(result)) {
       result += "_"
     }
```